akka-cluster-typed/src/test/scala/akka/cluster/ddata/typed/scaladsl/ReplicatorSpec.scala

```diff
@@ -34,59 +34,47 @@ object ReplicatorSpec {
   final case object Increment extends ClientCommand
   final case class GetValue(replyTo: ActorRef[Int]) extends ClientCommand
   final case class GetCachedValue(replyTo: ActorRef[Int]) extends ClientCommand
-  private sealed trait InternalMsg extends ClientCommand
-  private case class InternalUpdateResponse[A <: ReplicatedData](rsp: Replicator.UpdateResponse[A]) extends InternalMsg
-  private case class InternalGetResponse[A <: ReplicatedData](rsp: Replicator.GetResponse[A]) extends InternalMsg
-  private case class InternalChanged[A <: ReplicatedData](chg: Replicator.Changed[A]) extends InternalMsg
 
   val Key = GCounterKey("counter")
 
   def client(replicator: ActorRef[Replicator.Command])(implicit cluster: Cluster): Behavior[ClientCommand] =
     Actor.deferred[ClientCommand] { ctx ⇒
-      val updateResponseAdapter: ActorRef[Replicator.UpdateResponse[GCounter]] =
-        ctx.spawnAdapter(InternalUpdateResponse.apply)
-
-      val getResponseAdapter: ActorRef[Replicator.GetResponse[GCounter]] =
-        ctx.spawnAdapter(InternalGetResponse.apply)
-
-      val changedAdapter: ActorRef[Replicator.Changed[GCounter]] =
-        ctx.spawnAdapter(InternalChanged.apply)
-
-      replicator ! Replicator.Subscribe(Key, changedAdapter)
+      replicator ! Replicator.Subscribe(Key, ctx.self.upcast)
 
       def behavior(cachedValue: Int): Behavior[ClientCommand] = {
         Actor.immutable[ClientCommand] { (ctx, msg) ⇒
           msg match {
             case Increment ⇒
-              replicator ! Replicator.Update(Key, GCounter.empty, Replicator.WriteLocal, updateResponseAdapter)(_ + 1)
+              replicator ! Replicator.Update(Key, GCounter.empty, Replicator.WriteLocal, ctx.self.upcast)(_ + 1)
```

Comment:
we might want something named more clear than the `upcast` for this purpose, e.g. `ctx.selfAs`

Reply:
I find it weird it works given the type constraint on upcast, i imagine it wouldn't if you specified the type parameter explicitly